File: src/context_compiler/items.py
"""A single identifiable unit participating in a context bundle.

Minimal extraction of the historical item record: identity,
rendered content, and token accounting with explicit provenance.
`id` identifies this representation instance; `semantic_id`
optionally identifies the underlying information across
representations.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from context_compiler.tokens import estimate_tokens

SCHEMA_VERSION = "project_context.context_item.v1"

__all__ = ["SCHEMA_VERSION", "ContextItem", "estimate_tokens", "make_item"]


@dataclass(frozen=True)
class ContextItem:
    id: str
    source: str
    kind: str
    content: str
    position: int = 0
    token_count: int = 0
    token_provenance: str = "approximation"
    authority: str | None = None
    scope: str | None = None
    observed_at: str | None = None
    semantic_id: str | None = None
    ref: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": SCHEMA_VERSION,
            "id": self.id,
            "source": self.source,
            "kind": self.kind,
            "content": self.content,
            "position": self.position,
            "token_count": self.token_count,
            "token_provenance": self.token_provenance,
            "authority": self.authority,
            "scope": self.scope,
            "observed_at": self.observed_at,
            "semantic_id": self.semantic_id,
            "ref": self.ref,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ContextItem":
        version = data.get("schema_version", SCHEMA_VERSION)
        if version != SCHEMA_VERSION:
            raise ValueError(f"unsupported ContextItem schema: {version!r}")
        return cls(
            id=data["id"],
            source=data["source"],
            kind=data["kind"],
            content=data["content"],
            position=data.get("position", 0),
            token_count=data.get("token_count", 0),
            token_provenance=data.get("token_provenance", "approximation"),
            authority=data.get("authority"),
            scope=data.get("scope"),
            observed_at=data.get("observed_at"),
            semantic_id=data.get("semantic_id"),
            ref=data.get("ref"),
        )


def make_item(
    *,
    id: str,
    source: str,
    kind: str,
    content: str,
    authority: str | None = None,
    scope: str | None = None,
    semantic_id: str | None = None,
) -> ContextItem:
    """Build an item with a local token estimate. Position is assigned by
    the bundle builder, not here."""
    count, provenance = estimate_tokens(content)
    return ContextItem(
        id=id,
        source=source,
        kind=kind,
        content=content,
        token_count=count,
        token_provenance=provenance,
        authority=authority,
        scope=scope,
        semantic_id=semantic_id,
    )

File: src/core/items.ts
/**
 * ContextItem record (v1 schema preserved). Minimal port of the
 * historical item primitive: identity, rendered content, and token
 * accounting with explicit provenance.
 */

import { estimateTokens } from "./tokens.ts";

export const ITEM_SCHEMA = "project_context.context_item.v1";

export interface ContextItem {
  readonly id: string;
  readonly source: string;
  readonly kind: string;
  readonly content: string;
  readonly position: number;
  readonly tokenCount: number;
  readonly tokenProvenance: string;
  readonly authority: string | null;
  readonly scope: string | null;
  readonly observedAt: string | null;
  readonly semanticId: string | null;
  readonly ref: string | null;
}

export function itemToJSON(item: ContextItem): Record<string, unknown> {
  return {
    schema_version: ITEM_SCHEMA,
    id: item.id,
    source: item.source,
    kind: item.kind,
    content: item.content,
    position: item.position,
    token_count: item.tokenCount,
    token_provenance: item.tokenProvenance,
    authority: item.authority,
    scope: item.scope,
    observed_at: item.observedAt,
    semantic_id: item.semanticId,
    ref: item.ref,
  };
}

function reqStr(data: Record<string, unknown>, key: string): string {
  const value = data[key];
  if (typeof value !== "string") throw new Error(`invalid ${key}`);
  return value;
}

function optStrNull(data: Record<string, unknown>, key: string): string | null {
  const value = data[key] ?? null;
  if (value !== null && typeof value !== "string") {
    throw new Error(`invalid ${key}`);
  }
  return value;
}

export function itemFromJSON(data: unknown): ContextItem {
  if (typeof data !== "object" || data === null) {
    throw new Error("item is not an object");
  }
  const raw = data as Record<string, unknown>;
  const version = (raw["schema_version"] ?? ITEM_SCHEMA) as string;
  if (version !== ITEM_SCHEMA) {
    throw new Error(
      `unsupported ContextItem schema: ${JSON.stringify(version)}`,
    );
  }
  const position = raw["position"] ?? 0;
  const tokenCount = raw["token_count"] ?? 0;
  if (typeof position !== "number" || typeof tokenCount !== "number") {
    throw new Error("invalid item numbers");
  }
  return {
    id: reqStr(raw, "id"),
    source: reqStr(raw, "source"),
    kind: reqStr(raw, "kind"),
    content: reqStr(raw, "content"),
    position,
    tokenCount,
    tokenProvenance:
      typeof raw["token_provenance"] === "string"
        ? (raw["token_provenance"] as string)
        : "approximation",
    authority: optStrNull(raw, "authority"),
    scope: optStrNull(raw, "scope"),
    observedAt: optStrNull(raw, "observed_at"),
    semanticId: optStrNull(raw, "semantic_id"),
    ref: optStrNull(raw, "ref"),
  };
}

export function makeItem(input: {
  id: string;
  source: string;
  kind: string;
  content: string;
  authority?: string | null;
  scope?: string | null;
  semanticId?: string | null;
}): ContextItem {
  const [count, provenance] = estimateTokens(input.content);
  return {
    id: input.id,
    source: input.source,
    kind: input.kind,
    content: input.content,
    position: 0,
    tokenCount: count,
    tokenProvenance: provenance,
    authority: input.authority ?? null,
    scope: input.scope ?? null,
    observedAt: null,
    semanticId: input.semanticId ?? null,
    ref: null,
  };
}
